File: src/Hubs/Filehubs/FileHubs.cs
using Microsoft.AspNetCore.SignalR;
using OSManager.Service.FileService;
using System.IO;

namespace OSManager.Hubs.Filehubs;

public class FileHubs:Hub
{
    private readonly IFileService _fileService;

    public FileHubs(IFileService fileService)
    {
        _fileService = fileService;
    }

    [HubMethodName("CreateFiles")]
    public async Task CreateFiles(string relativePath)
    {
        if (string.IsNullOrWhiteSpace(relativePath))
        {
            await Clients.Caller.SendAsync("Error", "file name cannot be empty.");
            return;
        }
        try
        {
            var created = await _fileService.CreateFileAsync(relativePath);
            var fileInfo = new { Name = Path.GetFileName(created), Path = created };
            await Clients.Caller.SendAsync("CreateFiles", fileInfo);
        }
        catch (Exception ex)
        {
            await Clients.Caller.SendAsync("FileCreateError", ex.Message);
        }
    }

    [HubMethodName("UpdateFiles")]
    public async Task UpdateFiles(string relativePath, string newContent)
    {
        if (string.IsNullOrWhiteSpace(relativePath))
        {
            await Clients.Caller.SendAsync("Error", "file name cannot be empty.");
            return;
        }
        try
        {
            var file =  _fileService.UpdateFileContentAsync(relativePath, newContent);
            var fileInfo = new { Name = relativePath, Path = relativePath };
            await Clients.Caller.SendAsync("UpdateFiles", fileInfo);
        }
        catch (Exception ex)
        {
            await Clients.Caller.SendAsync("FileUpdateError", ex.Message);
        }
    }
}

File: ClientApp/src/Hubs/FileHubs/FileHubs.ts
import {fileConnection} from "../connection.ts";

export const FileCreateHubs = async (
    relativePath: string,
    onSuccess?: (fileInfo: { Name: string; Path: string }) => void,
    onError?: (error: string) => void
): Promise<void> => {
    if (!relativePath || relativePath.trim() === '') {
        throw new Error("Folder path cannot be empty");
    }
    const connection = fileConnection(); 

    try {
        if (onSuccess) {
            connection.on("CreateFiles", onSuccess);
        }
        if (onError) {
            connection.on("FileCreateError", onError);
        }
        await connection.start();
        await connection.invoke("CreateFiles", relativePath);
    } catch (error) {
        throw error;
    } finally {
        if (onSuccess) {
            connection.off("CreateFiles", onSuccess);
        }
        if (onError) {
            connection.off("FileCreateError", onError);
        }
        await connection.stop();
    }
};

export const FileUpdateHubs = async (
    relativePath: string,
    newContent: string,
    onSuccess?: (fileInfo: { Name: string; Path: string }) => void,
    onError?: (error: string) => void
): Promise<void> => {
    if (!relativePath || relativePath.trim() === '') {
        throw new Error("Folder path cannot be empty");
    }
    const connection = fileConnection();

    try {
        if (onSuccess) {
            connection.on("UpdateFiles", onSuccess);
        }
        if (onError) {
            connection.on("FileUpdateError", onError);
        }
        await connection.start();
        await connection.invoke("UpdateFiles", relativePath, newContent);
    } catch (error) {
        throw error;
    } finally {
        if (onSuccess) {
            connection.off("UpdateFiles", onSuccess);
        }
        if (onError) {
            connection.off("FileUpdateError", onError);
        }
        await connection.stop();
    }
};